src/org/sosy_lab/java_smt/basicimpl/AbstractProver.java

```diff
@@ -47,7 +54,13 @@ public abstract class AbstractProver<T> implements BasicProverEnvironment<T> {
 
   private static final String TEMPLATE = "Please set the prover option %s.";
 
-  protected AbstractProver(Set<ProverOptions> pOptions) {
+  protected final @Nullable ShutdownNotifier proverShutdownNotifier;
```

Comment:
Why make it `@Nullable` instead of using a dummy `ShutdownNotifier`? Do you think the additional volatile read on every `shutdownIfNecessary()` is too expensive compared to the null check for users who do not need this feature?

And why are the `ShutdownNotifier`s protected? Why should subclasses access them and not just `shutdownIfNecessary()`?

Reply:
No, its `null` because i need a way of discerning if we have a usable prover based notifier or not, as some solvers do not support this. If there were a check `isDummy()` i could use a dummy.

Reply:
But why and where is this information actually needed?

Reply:
I don't remember the reason why currently. I know that i tried it and found some problem with it.

With all the new changes this seems to work now though! Which obviously makes better code, and i think we can take the minimal overhead.